arch/x86/entry.S

```diff
@@ -32,70 +32,86 @@
 #include <usermode.h>
 #include <errno.h>
 
-.macro _handle_usermode cr3
+.macro MASK_USER_FLAGS
+    PUSHF
+    andl $~(USERMODE_FLAGS_MASK), (%_ASM_SP)
+    POPF
+.endm
+
+.macro _from_usermode switch_stack=0
+    SET_CR3 cr3
+    swapgs
+    .if \switch_stack == 1
+        SWITCH_STACK
+    .endif
+.endm
+
+.macro _to_usermode switch_stack=0
+    .if \switch_stack == 1
+        SWITCH_STACK
+    .endif
+    swapgs
+    SET_CR3 user_cr3
+.endm
+
+.macro cond_from_usermode
     testb $0x3, cpu_exc_cs(%_ASM_SP)
     jz 1f /* skip if from kernel mode */
-        SET_CR3 \cr3
-        swapgs
+        MASK_USER_FLAGS
+        _from_usermode
     1:
 .endm
 
-.macro enter_from_usermode
-    _handle_usermode cr3
+.macro cond_to_usermode
+    testb $0x3, cpu_exc_cs(%_ASM_SP)
+    jz 1f /* skip if from kernel mode */
+        _to_usermode
+    1:
 .endm
 
-.macro enter_to_usermode
-    _handle_usermode user_cr3
+.macro syscall_from_usermode switch_stack=1
+    _from_usermode switch_stack=\switch_stack
 .endm
 
-.macro syscall_from_usermode
-    SET_CR3 cr3
-    swapgs
-    SWITCH_STACK
-.endm
-
-.macro syscall_to_usermode
-    SWITCH_STACK
-    swapgs
-    SET_CR3 user_cr3
+.macro syscall_to_usermode switch_stack=1
+    _to_usermode switch_stack=\switch_stack
 .endm
 
 .macro exception_handler sym vec has_error_code
 ENTRY(entry_\sym)
-    enter_from_usermode
-
     .if \has_error_code == 0
         push $0
     .endif
-
-#if defined(__x86_64__)
-    movl $\vec, cpu_exc_vector(%_ASM_SP)
-#else
     push $\vec
-#endif
 
+    cond_from_usermode
     jmp handle_exception
 END_FUNC(entry_\sym)
 .endm
 
-.macro MASK_USER_FLAGS
-    PUSHF
-    andl $~(USERMODE_FLAGS_MASK), (%_ASM_SP)
-    POPF
+.macro check_syscall_exit
+    cmp $SYSCALL_EXIT, %_ASM_AX
+    jnz 1f
+        _from_usermode
+        jmp syscall_exit
+    1:
 .endm
 
-.macro interrupt_handler sym func
+.macro interrupt_handler sym func vec
 ENTRY(asm_interrupt_handler_\sym)
-    enter_from_usermode
-
-    MASK_USER_FLAGS
+    push $0 /* fake error_code */
+    push $\vec /* IRQ vector */
+    cond_from_usermode
 
     cld
     SAVE_ALL_REGS
     call \func
```

Comment:
Hmm. It looks like the IRQ handler doesn't care about the current register state nor the exception frame. So we could simplify it further by omitting the needles `push $0; push \vec` (sorry, I was assuming the exception frame gets passed on to the C handler but, apparently, it isn't).

So what about changing `cond_from_usermode` to take an offset for the cs access?:

```asm
.macro cond_to_usermode offset=cpu_exc_cs
    testb $0x3, \offset(%_ASM_SP)
    jz 1f /* skip if from kernel mode */
        _to_usermode
    1:
.endm
```
and use it like this here?:
```asm
cond_from_usermode offset=(cpu_exc_cs-cpu_exc_ip)
```

Then we can omit the two pushes above and the add below.

Reply:
Actually, I would like to keep it as is, because I am planning to pass the `cpu_regs_t` to the interrupt handlers in the future (for example for task switching purposes etc). Also having the IRQ vector stored might be useful for future enhancements to the interrupt handlers.